c/rc_lock.c

```diff
@@ -0,0 +1,462 @@
+// uncomment to enable printf in CKB-VM
+// #define CKB_C_STDLIB_PRINTF
+
+// it's used by blockchain-api2.h, the behavior when panic
+#ifndef MOL2_EXIT
+#define MOL2_EXIT ckb_exit
+#endif
+int ckb_exit(signed char);
+
+// clang-format off
+#include <stdio.h>
+#include "blockchain-api2.h"
+#include "blockchain.h"
+#include "ckb_consts.h"
+#include "blake2b.h"
+#if defined(CKB_USE_SIM)
+#include "ckb_syscall_rc_lock_sim.h"
+#else
+#include "ckb_syscalls.h"
+#endif
+
+#include "ckb_smt.h"
+#include "secp256k1_helper.h"
+// CHECK is defined in secp256k1
+#undef CHECK
+#include "rce.h"
+#include "rc_lock_mol2.h"
+// clang-format on
+
+#define BLAKE2B_BLOCK_SIZE 32
+#define BLAKE160_SIZE 20
+#define PUBKEY_SIZE 33
+#define TEMP_SIZE 32768
+#define RECID_INDEX 64
+/* 32 KB */
+#define MAX_WITNESS_SIZE 32768
+#define SCRIPT_SIZE 32768
+#define SIGNATURE_SIZE 65
+#define ONE_BATCH_SIZE 32768
+#define MAX_LOCK_SCRIPT_HASH_COUNT 2048
+
+enum RcLockErrorCode {
+  ERROR_SECP_RECOVER_PUBKEY = -11,
+  ERROR_SECP_PARSE_SIGNATURE = -14,
+  ERROR_SECP_SERIALIZE_PUBKEY = -15,
+  ERROR_PUBKEY_BLAKE160_HASH = -31,
+  // rc lock error code is starting from 80
+  ERROR_UNKNOWN_FLAGS = 80,
+  ERROR_LOCK_SCRIPT_HASH_NOT_FOUND,
+  ERROR_PROOF_LENGTH_MISMATCHED,
+  ERROR_NO_RCRULE,
+  ERROR_NO_WHITE_LIST,
+};
+
+enum IdentityFlagsType {
+  IdentityFlagsPubkeyHash = 0,
+  IdentityFlagsOwnerLock = 1,
+  IdentityFlagsPubkeyHashRc = 2,
+  IdentityFlagsOwnerLockRc = 3,
+};
+
+typedef struct IdentityType {
+  // IdentityFlagsType
+  uint8_t flags;
+  // blake160 (20 bytes) hash of lock script or pubkey
+  uint8_t blake160[20];
+} IdentityType;
+
+typedef struct ArgsType {
+  IdentityType id;
+  uint8_t rc_root[32];
+} ArgsType;
+
+// make compiler happy
+int make_cursor_from_witness(WitnessArgsType *witness, bool *_input) {
+  return 0;
```

Comment:
It's better if we return an error code if the sole purpose here is for compiler reason